modelzoo/DBMTL/train.py

```diff
@@ -0,0 +1,709 @@
+    if scale < 0.:
+      raise ValueError(f'Setting a scale less than 0 on a regularizer: {scale}.')
+    if scale == 0.:
+      return lambda _: None
+
+  def l2(weights):
+    with tf.name_scope(scope, 'l2_regularizer', [weights]) as name:
+      my_scale = tf.convert_to_tensor(scale, dtype=weights.dtype.base_dtype, name='scale')
+      return tf.math.multiply(my_scale, tf.nn.l2_loss(weights), name=name)
+
+  return l2
+
+class DBMTL():
+    def __init__(self,
+                 input,
+                 feature_column,
+                 bottom_dnn,
+                 towers,
+                 relation_dnn,
+                 l2_scale,
+                 learning_rate=0.1,
+                 bf16=False,
+                 input_layer_partitioner=None,
+                 dense_layer_partitioner=None):
+        self.__feature_column = feature_column
+        self.__bottom_dnn = bottom_dnn
+        self.__towers = towers
+        self.__relation_dnn = relation_dnn
+        self.__l2_regularization = l2_regularizer(l2_scale) if l2_scale else None
+        self.__learning_rate = learning_rate
+        self.__bf16 = bf16
+        self.__input_layer_partitioner = input_layer_partitioner
+        self.__dense_layer_partitioner = dense_layer_partitioner
+
+        self.feature = input[0]
+        self.label = input[1]
+
+        self.model = self.__build_model()
+        
+        with tf.name_scope('head'):
+            self.train_op, self.loss = self.compute_loss()
+            self.acc, self.acc_op = tf.metrics.accuracy(labels=self.label, predictions=tf.round(self.model))
+            self.auc, self.auc_op = tf.metrics.auc(labels=self.label, predictions=self.model, num_thresholds=1000)
+
+            tf.summary.scalar('eval_acc', self.acc)
+            tf.summary.scalar('eval_auc', self.auc)
+    
+    def compute_loss(self):
+        bce_loss_func = tf.keras.losses.BinaryCrossentropy()
+        self.model = tf.squeeze(self.model)
+        loss = tf.math.reduce_mean(bce_loss_func(self.label, self.model))
+        tf.summary.scalar('loss', loss)
+
+        # TODO: add decay
+        self.global_step = tf.train.get_or_create_global_step()
+        optimizer = tf.train.GradientDescentOptimizer(learning_rate=self.__learning_rate)
+
+        train_op = optimizer.minimize(loss, global_step=self.global_step)
+
+        return train_op, loss
+    
+    def __build_model(self):
+        TAG_COLUMN = ['tag_category_list', 'tag_brand_list']
+        for key in TAG_COLUMN:
+            self.feature[key] = tf.strings.split(self.feature[key], '|')
+
+        with tf.variable_scope('input_layer',
+                               partitioner=self.__input_layer_partitioner,
+                               reuse=tf.AUTO_REUSE):
+            input_emb = tf.feature_column.input_layer(self.feature,
+                                                  self.__feature_column)
+        # Shared Layer
+        with tf.variable_scope(f'bottom_dnn_'):
```

Comment:
fyi, You don't need to use f-string formatting if you are not passing any variable to the string.

Reply:
Changed